### PR TITLE
No blank backend role before adding a new one in Create User page

### LINKED ISSUE #1307: [BUG] Backend roles section of Create User page allows blank backend role
**What is the bug?**

The user creation screen allows a user to continue to add backend roles even if no text is entered. In addition to creating blank roles, the remove button is always enabled even if no text is entered. If there is only one backend role, the remove button will blank out the text that has been filled. If there is more than one role then remove will remove the entire row. 

Screen Shot 2023-01-23 at 10 24 00 AM

Blank backend roles do show up for the logged in user with a blank backend role:

Screen Shot 2023-01-23 at 10 27 41 AM

**How can one reproduce the bug?**
1. Go to Security -> Internal Users
2. Create a User
3. Click on `Add another backend role`

**What is the expected behavior?**

The backend role should not be blank before a new one can be added

Tested on 2.5.0

### PR DESCRIPTION
**Pull request by nurSaadat**
### Description
Perform a check on the list of backend roles whether it contains an empty value. If it does, display an error message and don't allow the user to create additional backend roles until the existing blank panels are filled with values.

### Category
[Bug fix]

### Why these changes are required?
Those changes help to make UX more intuitive and also prevent undesired user behaviour.

### What is the old behavior before changes and new behavior after changes?
Before the changes, new empty backend roles were added every time the "Add another backend role" button was clicked.
After the change, it is checked whether there are any empty fields before adding a new one.

### Issues Resolved
* Resolve: https://github.com/opensearch-project/security-dashboards-plugin/issues/1307

### Testing
- First case : the user tries to add a new backend role while the last text field is empty.
image
image

- Second case: the user tries to add a new backend role while any previously created text field is empty
image

### Check List
- [x] New functionality includes testing
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by RyanL1997 — COMMENTED** [review]
Hi @nurSaadat, great job on fixing this! There are only couple problems we need to address before we merge this fix:
1. We need to modify/add unit test for this fix, but still it is nice to see you have tested it manually :) I believe the place for adding test case is [here](https://github.com/opensearch-project/security-dashboards-plugin/blob/b3b3120144e50d229bb09aec50991a383df6dafb/public/apps/configuration/panels/internal-user-edit/test/backend-role-panel.test.tsx). 
2. Fix the lint, and you can find out more details [here](https://github.com/opensearch-project/security-dashboards-plugin/actions/runs/4494901551/jobs/7908143651?pr=1384).
3. I also left some comments on the code:

File: public/apps/configuration/panels/internal-user-edit/backend-role-panel.tsx
Comment (RyanL1997) [inline public/apps/configuration/panels/internal-user-edit/backend-role-panel.tsx:52]: Should use `===`, instead of `==`.

File: public/apps/configuration/panels/internal-user-edit/backend-role-panel.tsx
Comment (RyanL1997) [inline public/apps/configuration/panels/internal-user-edit/backend-role-panel.tsx:79]: Unrelated change, please revert.

File: public/apps/configuration/panels/internal-user-edit/backend-role-panel.tsx
Comment (RyanL1997) [inline public/apps/configuration/panels/internal-user-edit/backend-role-panel.tsx:112]: Same with the above, should use `!==` instead of `==`.

**Review by scrawfor99 — DISMISSED** [review]
Hi @nurSaadat, great job fixing this! Thank you for your contribution. Can you confirm that you checked that the issue from before is fixed? I know frontend testing can be a pain, so I won't request that you write a whole cypress test, but I just want to check with you that it did seem to fix the issue. 

Edit: @peternied has a good point so I rescind my comment. If you could add a test to confirm this is as expected that would be great.

**Comment by peternied** [comment]
> I know frontend testing can be a pain, so I won't request that you write a whole cypress test, but I just want to check with you that it did seem to fix the issue.

@scrawfor99 I disagree, we should always have a tests associated with changes.  However, if it is a unit test, snapshot, jest, or cypress should be evaluated by maintainers

**Comment by nurSaadat** [comment]
@peternied I don't quite get what is wrong with my tests, so they fail.

Given that I have changed the contents of public/apps/configuration/panels/internal-user-edit/backend-role-panel.tsx, so that EuiFormRow has additional props `error`, `isInvalid`, and EuiFieldText has `isInvalid`
```
<EuiFormRow
              label={arrayIndex === 0 ? 'Backend role' : ''}
              error={roleEmptyErrorMessage}
              isInvalid={arrayIndex === emptyRoleIndex && !isEmpty(roleEmptyErrorMessage)}
            >
              <EuiFieldText
                isInvalid={arrayIndex === emptyRoleIndex && !isEmpty(roleEmptyErrorMessage)}
                id={`backend-role-${arrayIndex}`}
                value={backendRole}
                onChange={(e) => {
                  updateElementInArrayHandler(setBackendRoles, [arrayIndex])(e.target.value);
                  setRoleEmptyErrorMessage('');
                }}
                placeholder="Type in backend role"
              />
            </EuiFormRow>
       ...

      <EuiButton
        id="backend-role-add-row"
        onClick={() => {
          if (state.indexOf('') !== -1) {
            setRoleEmptyErrorMessage('Type a backend role before adding a new one');
            setEmptyRoleIndex(state.indexOf(''));
          } else {
            setRoleEmptyErrorMessage('');
            appendElementToArray(setState, [], '');
          }
        }}
      >
        Add another backend role
      </EuiButton>
```
I don't understand why following tests fail:
1. add backend role when the previous role is blank. Expected behavior: call `AppendElementToArray` once. But now it calls function `appendElementToArray` twice.
```
    it('add backend role when the previous role is blank', () => {
      const component = shallow(<BackendRolePanel state={sampleState} setState={setState} />);
      component.find('#backend-role-add-row').simulate('click');
      expect(appendElementToArray).toBeCalledWith(setState, [], '');
      component.find('#backend-role-add-row').simulate('click');
      expect(appendElementToArray).not.toBeCalledWith(setState, [], '');
    });
```
2. add backend role when one of the previous roles is blank. Expected behavior: set `EuiFormRow` prop `isInvalid` to `true` and set `error` message as 'Type a backend role before adding a new one'.
```
    it('add backend role when one of the previous roles is blank', () => {
      const component = shallow(<BackendRolePanel state={sampleState} setState={setState} />);
      component.find('#backend-role-0').simulate('change', { target: { value: '' } });
      component.find('#backend-role-add-row').simulate('click');
      expect(component.find(EuiFormRow).at(0).prop('error')).toBe('Type a backend role before adding a new one');
      expect(component.find(EuiFormRow).at(0).prop('isInvalid')).toBe(true);
    });
```
My confidence with Jest is still building up. Could you please suggest me what I am doing wrong?

**Comment by peternied** [comment]
@nurSaadat Rather than a jest-based integration test, try a snapshot test.  For this change a jest integration is overkill as they are complex to build and maintain.

There are a number of 'snapshot tests' that verify the rendering of a component's DOM, which I think might be a better fit for this change. There is a component `permissions-tree` that has a snapshot test that might be a good starting point, links [[component]](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/public/apps/configuration/panels/permission-tree/permission-tree.tsx), [[test]](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/public/apps/configuration/panels/permission-tree/test/permission-tree.test.tsx), [[snapshot]](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/public/apps/configuration/panels/permission-tree/test/__snapshots__/permission-tree.test.tsx.snap).

Please post an update to the pull request with your test(s) even if it does not work, this way others can provide feedback.

**Comment by peternied** [comment]
@nurSaadat I am not going to be able to follow the code in your comment and provide you useful info feedback.  Please add files to the pull request and annotation them with comments where you include the text of error output or unexpected behavior.

Pull requests can be messy while they are coming together, you can always iterate on them or include/remove files - thanks for digging in on this!

**Review by peternied — COMMENTED** [review]
@nurSaadat Thanks for adding the tests, looks like the linters are failing that is preventing the tests from running.

Other than the linting issues, the new tests look good to me

**Review by DarshitChanpura — DISMISSED** [review]
LGTM! Thank you @nurSaadat. Once the linter errors are fixed, this is good to go!

**Review by cwperks — APPROVED** [review]

**Review by RyanL1997 — APPROVED** [review]

**Pull request merged**
